# NLP Project

## Final project implement a transformer from scratch

### steps of the final project :
- Import necessary libraries and modules
- Define the basic building blocks: Multi-Head Attention, Position-wise Feed-Forward Networks, Positional Encoding
- Build the Encoder and Decoder layers
- Combine Encoder and Decoder layers to create the complete Transformer model
- Prepare your data [ clean and toknize and do all the cleaning steps nessacry from your point of veiw]
- Train the model
- test the model 


## Importing Libararies Needed

In [89]:
# for data reading and slicing
import numpy as np
import pandas as pd

# for string manipulate
import string
import re

# for handling contractions
import contractions

# Import Natural language Toolkit for stopwords and tokization and stemming and lemmatization
import nltk

# Import Natural language Toolkit for stopwords 
from nltk.corpus import stopwords

# Import Natural language Toolkit for tokization 
from nltk.tokenize import word_tokenize

# Import Natural language Toolkit for lemmatization 
from nltk.stem import WordNetLemmatizer

# for train and test split 
from sklearn.model_selection import train_test_split

# for randomly selecting the data
import random

# for transformer Neural Networks inside it using pytorch
import torch

# for pytorch neural network
import torch.nn as nn

# for pytorch optimizer
import torch.optim as optim

# for pytorch convert data to tensor and create train and test data
from torch.utils.data import DataLoader, TensorDataset

# for pytorch activation ,loss functions
import torch.nn.functional as F

# for evaluating the model 
from sklearn.metrics import classification_report

In [27]:
# Download NLTK tokenizer data
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [28]:
# Download NLTK Stop Words data
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [29]:
# Download the WordNet dataset
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

## Reading and Analyzing and Understanding the data

In [30]:
# Read CSV 
df = pd.read_csv(r"C:\Users\User\Desktop\koleya\summer 2\SPRINTS  AI & ML program\Assigment\assigment 6\train.csv")

# look inside the first 10 rows of data
df.head()

id                                       comment_text  toxic  \
0  0000997932d777bf  Explanation\nWhy the edits made under my usern...      0   
1  000103f0d9cfb60f  D'aww! He matches this background colour I'm s...      0   
2  000113f07ec002fd  Hey man, I'm really not trying to edit war. It...      0   
3  0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...      0   
4  0001d958c54c6e35  You, sir, are my hero. Any chance you remember...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0  
3             0        0       0       0              0  
4             0        0       0       0              0

In [31]:
# get shape of data
df.shape 

(159571, 8)

#### There is 159571 rows and 8 columns

### Information of each column

In [32]:
# get info of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             159571 non-null  object
 1   comment_text   159571 non-null  object
 2   toxic          159571 non-null  int64 
 3   severe_toxic   159571 non-null  int64 
 4   obscene        159571 non-null  int64 
 5   threat         159571 non-null  int64 
 6   insult         159571 non-null  int64 
 7   identity_hate  159571 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 9.7+ MB


#### There is no nulls in data and all columns are int except columns : id , comment_text are string

### checking for nulls

In [33]:
# get the number of nulls in each column
df.isna().sum()

id               0
comment_text     0
toxic            0
severe_toxic     0
obscene          0
threat           0
insult           0
identity_hate    0
dtype: int64

#### there is no nulls in the data

### Describtion of data

In [34]:
# get the describtion of statistics of the int columns
df.describe().T

count      mean       std  min  25%  50%  75%  max
toxic          159571.0  0.095844  0.294379  0.0  0.0  0.0  0.0  1.0
severe_toxic   159571.0  0.009996  0.099477  0.0  0.0  0.0  0.0  1.0
obscene        159571.0  0.052948  0.223931  0.0  0.0  0.0  0.0  1.0
threat         159571.0  0.002996  0.054650  0.0  0.0  0.0  0.0  1.0
insult         159571.0  0.049364  0.216627  0.0  0.0  0.0  0.0  1.0
identity_hate  159571.0  0.008805  0.093420  0.0  0.0  0.0  0.0  1.0

#### All int columns are binary 

### Types of columns

In [35]:
# get types of each column
print(df.dtypes)

id               object
comment_text     object
toxic             int64
severe_toxic      int64
obscene           int64
threat            int64
insult            int64
identity_hate     int64
dtype: object


#### All columns are int except ID and comment_text

### Drop Duplicates in column id and comment_text

In [36]:
# drop identical rows in column comment_text
df.drop_duplicates(subset='comment_text',inplace=True) 

# drop identical rows in column id
df.drop_duplicates(subset='id',inplace=True) 

In [37]:
df.shape 

(159571, 8)

#### There is no duplicantes in columns : comment_text , id ; because number of rows is the same before drop duplicants and  after drop duplicants

### See unique values in data

In [38]:
# get the number of unique value of each column
df.nunique()

id               159571
comment_text     159571
toxic                 2
severe_toxic          2
obscene               2
threat                2
insult                2
identity_hate         2
dtype: int64

#### All rows in columns : id , comment_text has a unique value

### See what is the unique values in each int column

In [39]:
# see each unique value in toxic quantity of it
df['toxic'].value_counts()

toxic
0    144277
1     15294
Name: count, dtype: int64

In [40]:
# see each unique value in severe_toxic quantity of it
df['severe_toxic'].value_counts()

severe_toxic
0    157976
1      1595
Name: count, dtype: int64

In [41]:
# see each unique value in obscene quantity of it
df['obscene'].value_counts()

obscene
0    151122
1      8449
Name: count, dtype: int64

In [42]:
# see each unique value in threat quantity of it
df['threat'].value_counts()

threat
0    159093
1       478
Name: count, dtype: int64

In [43]:
# see each unique value in insult quantity of it
df['insult'].value_counts()

insult
0    151694
1      7877
Name: count, dtype: int64

In [44]:
# see each unique value in identity_hate quantity of it
df['identity_hate'].value_counts()

identity_hate
0    158166
1      1405
Name: count, dtype: int64

#### I can say from in the above cells that all rows has unique id and comment_text , and the rest columns have only zero or one

### See if is there rows that has inappropriate comment and how many of the data it is

In [45]:
# get the rows that is appropriate by making a condition that all int values = 0
condition = (df["toxic"]==0)&(df["severe_toxic"]==0)&(df["obscene"]==0)&(df["threat"]==0)&(df["insult"]==0)&(df["identity_hate"]==0)

df[condition]

id                                       comment_text  \
0       0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1       000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2       000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3       0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4       0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
...                  ...                                                ...   
159566  ffe987279560d7ff  ":::::And for the second time of asking, when ...   
159567  ffea4adeee384e90  You should be ashamed of yourself \n\nThat is ...   
159568  ffee36eab5c267c9  Spitzer \n\nUmm, theres no actual article for ...   
159569  fff125370e4aaaf3  And it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  "\nAnd ... I really don't think you understand...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[143346 rows x 8 columns]

In [46]:
# get the number of rows that meets the condition
num = df[condition].shape[0] 

# divide the number by whole length of data * 100 to get percentage
Percentage_of_clean_data = (num / len(df))*100 

# get the percentage of appropriate data
Percentage_of_toxic_data = 100 - Percentage_of_clean_data

print("Percentage of clean data {} % \nPercentage of toxic data {} %".format(Percentage_of_clean_data,Percentage_of_toxic_data))

Percentage of clean data 89.83211235124176 % 
Percentage of toxic data 10.167887648758239 %


#### About 90% of the data containes appropriate comments and 10 % is inappropriate comments

## Cleaning The Data

### steps of cleaning the data :
- lowercase the data : to ensure consistency
- Removing URLs
- Handling Contractions , for example change "can't" to "cannot"
- Removing Special Characters 
- Remove punctuation
- Remove Numbers
- stop words : to reduce the input's dimensionality

### Implementation of a function that counts the rows that have  ( URLs , Special Characters , Punctuation , Stop Words , Numbers )

In [47]:
def count_rows(column):
    # Initialize count = zero
    urls_count = 0
    special_chars_count = 0
    punctuation_count = 0
    stop_word_count = 0
    numeric_count = 0
    
    # make a set of all stop words
    stop_words = set(stopwords.words('english'))

    for text in column:
        # Count URLs
        # if URLs found
        if re.search(r'http\S+|www\S+|https\S+', text):
        # add by one
            urls_count += 1

        # Count special characters except english letters and spaces
        # if special characters found 
        if re.search(r'[^A-Za-z0-9\s]+', text):
        # add by one
            special_chars_count += 1
        
        # Count numbers
        # if number found 
        if  re.search(r'\d', text):
        # add by one
            numeric_count +=1

        # Count punctuation
        # if punctuation found 
        if any(char in string.punctuation for char in text):
        # add by one
            punctuation_count += 1

        # Count stop words
        words = text.split()
        # if the word is stop word found 
        if any(word.lower() in stop_words for word in words):
        # add by one
            stop_word_count += 1

    return urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count

### Implementation of a function that lowercase the data , Remove URLs , Return the word to it's basic form  , Remove Special Characters , Remove punctuation , Remove Numbers

In [48]:
def clean_data(text):

    # Lowercase the text
    text = text.lower()

    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)

    # Handle Contractions
    text = contractions.fix(text)
    
    # Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))
    
    # Remove any special characters except english letters and spaces 
    text=re.sub(r'[^A-Za-z\s]+',' ',text)
    
    # Remove stop words
    # Tokenize the text
    tokens = nltk.word_tokenize(text)
    # create a set of stop words using NLTK libarary
    stop_words = set(nltk.corpus.stopwords.words('english'))
    # iterate over the data and remove any stop word in it
    filtered_tokens = [word for word in tokens if word not in stop_words]
    # Join the filtered tokens back into a string
    text = ' '.join(filtered_tokens)
    
    return text

### count the number of rows that have URLs , Special Characters , punctuation , numbers , and stop words BEFORE clean data function 

In [49]:
# apply the counting count_rows_urls_and_special_chars_and_punctuation_and_stop_words_and_numbers functions
urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count= count_rows(df['comment_text'])

print(f"Number of rows with URLs: {urls_count}")
print(f"Number of rows with special characters: {special_chars_count}")
print(f"Number of rows with punctuation: {punctuation_count}")
print(f"Number of rows with stop words: {stop_word_count}")
print(f"Number of rows with Numbers: {numeric_count}")

Number of rows with URLs: 5151
Number of rows with special characters: 155146
Number of rows with punctuation: 155106
Number of rows with stop words: 154877
Number of rows with Numbers: 51212


### apply clean data function on data

In [50]:
# apply clean_data function on comment_text column
df['comment_text'] = df['comment_text'].apply(clean_data)

### count the number of rows that have URLs , Special Characters , punctuation , numbers , and stop words AFTER clean data function 

In [51]:
# apply the counting count_rows_urls_and_special_chars_and_punctuation_and_stop_words_and_numbers functions
urls_count, special_chars_count , punctuation_count , stop_word_count , numeric_count= count_rows(df['comment_text'])

print(f"Number of rows with URLs: {urls_count}")
print(f"Number of rows with special characters: {special_chars_count}")
print(f"Number of rows with punctuation: {punctuation_count}")
print(f"Number of rows with stop words: {stop_word_count}")
print(f"Number of rows with Numbers: {numeric_count}")

Number of rows with URLs: 0
Number of rows with special characters: 0
Number of rows with punctuation: 0
Number of rows with stop words: 0
Number of rows with Numbers: 0


#### The function successfully cleaned the data from URLs , Special Characters , punctuation , numbers , and stop words

In [52]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
159566  ffe987279560d7ff  second time asking view completely contradicts...   
159567  ffea4adeee384e90               ashamed horrible thing put talk page   
159568  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
159569  fff125370e4aaaf3  looks like actually put speedy first version d...   
159570  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

### Drop Duplicates After Cleaning

#### cleaning function will make some rows hold the same text , so will drop duplicantes on column comment text

#### Apply function on text data

In [53]:
# drop duplicates on comment text column
df.drop_duplicates(subset='comment_text',inplace=True) 

# reset index and drop column index that will be produced from reset index
df = df.reset_index(drop=True)

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157664  ffe987279560d7ff  second time asking view completely contradicts...   
157665  ffea4adeee384e90               ashamed horrible thing put talk page   
157666  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157667  fff125370e4aaaf3  looks like actually put speedy first version d...   
157668  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
157664      0             0        0       0       0              0  
157665      0             0        0       0       0              0  
157666      0             0        0       0       0              0  
157667      0             0        0       0       0              0  
157668      0             0        0       0       0              0  

[157669 rows x 8 columns]

#### See the number of rows has reduced from 159571 to 157669 , so there was duplicantes in comment_text column after cleaning

## Toknization

In [54]:
# function that tokenize the data which is built in 
def tokenize_text(text):
    tokens = word_tokenize(text)
    return tokens

### apply function on data

In [55]:
# apply function on mentioned column 
df['tokenized_column'] = df['comment_text'].apply(tokenize_text)

In [56]:
df.head(10)

id                                       comment_text  toxic  \
0  0000997932d777bf  explanation edits made username hardcore metal...      0   
1  000103f0d9cfb60f  daww matches background colour seemingly stuck...      0   
2  000113f07ec002fd  hey man really trying edit war guy constantly ...      0   
3  0001b41b1c6bb37e  make real suggestions improvement wondered sec...      0   
4  0001d958c54c6e35                      sir hero chance remember page      0   
5  00025465d4725e87           congratulations well use tools well talk      0   
6  0002bcb3da6cb337                        cocksucker piss around work      1   
7  00031b1e95af7921  vandalism matt shirvington article reverted pl...      0   
8  00037261f536c51d  sorry word nonsense offensive anyway intending...      0   
9  00040093b2687caa               alignment subject contrary dulithgow      0   

   severe_toxic  obscene  threat  insult  identity_hate  \
0             0        0       0       0              0   
1             0        0       0       0              0   
2             0        0       0       0              0   
3             0        0       0       0              0   
4             0        0       0       0              0   
5             0        0       0       0              0   
6             1        1       0       1              0   
7             0        0       0       0              0   
8             0        0       0       0              0   
9             0        0       0       0              0   

                                    tokenized_column  
0  [explanation, edits, made, username, hardcore,...  
1  [daww, matches, background, colour, seemingly,...  
2  [hey, man, really, trying, edit, war, guy, con...  
3  [make, real, suggestions, improvement, wondere...  
4                [sir, hero, chance, remember, page]  
5    [congratulations, well, use, tools, well, talk]  
6                   [cocksucker, piss, around, work]  
7  [vandalism, matt, shirvington, article, revert...  
8  [sorry, word, nonsense, offensive, anyway, int...  
9          [alignment, subject, contrary, dulithgow]

### Made a new column containing the tokenized data

## Lemmatization 

### To return the word to it's basic form

### implementation of Lemmatization function 

In [57]:
def lemmatize_tokens(tokens):
    # initialize WordNetLemmatizer
    lemmatizer = WordNetLemmatizer()

    # Lemmatize each word in the list of tokens
    lemmatized_tokens = [lemmatizer.lemmatize(word) for word in tokens]

    # make the words into sentence instead of a list
    return ' '.join(lemmatized_tokens)

### apply function on tokenized column 

In [58]:
# apply function on tokenized column
df['lemmatization_column']=df['tokenized_column'].apply(lemmatize_tokens)

In [59]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157664  ffe987279560d7ff  second time asking view completely contradicts...   
157665  ffea4adeee384e90               ashamed horrible thing put talk page   
157666  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157667  fff125370e4aaaf3  looks like actually put speedy first version d...   
157668  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157664      0             0        0       0       0              0   
157665      0             0        0       0       0              0   
157666      0             0        0       0       0              0   
157667      0             0        0       0       0              0   
157668      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157664  [second, time, asking, view, completely, contr...   
157665        [ashamed, horrible, thing, put, talk, page]   
157666  [spitzer, umm, actual, article, prostitution, ...   
157667  [looks, like, actually, put, speedy, first, ve...   
157668  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  
0       explanation edits made username hardcore metal...  
1       daww match background colour seemingly stuck t...  
2       hey man really trying edit war guy constantly ...  
3       make real suggestion improvement wondered sect...  
4                           sir hero chance remember page  
...                                                   ...  
157664  second time asking view completely contradicts...  
157665               ashamed horrible thing put talk page  
157666  spitzer umm actual article prostitution ring c...  
157667  look like actually put speedy first version de...  
157668  really think understand came idea bad right aw...  

[157669 rows x 10 columns]

### Made a new column containing the lemmatization data

### Drop Duplicates After lemmatization

#### Lemmatization function will make some rows hold the same text , so will drop duplicantes on column lemmatization

#### Apply drop duplicates on data

In [60]:
# drop duplicates on lemmatization column
df.drop_duplicates(subset='lemmatization_column',inplace=True) 

In [61]:
# reset index and drop column index that will be produced from reset index
df = df.reset_index(drop=True)

In [62]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157627  ffe987279560d7ff  second time asking view completely contradicts...   
157628  ffea4adeee384e90               ashamed horrible thing put talk page   
157629  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157630  fff125370e4aaaf3  looks like actually put speedy first version d...   
157631  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157627      0             0        0       0       0              0   
157628      0             0        0       0       0              0   
157629      0             0        0       0       0              0   
157630      0             0        0       0       0              0   
157631      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157627  [second, time, asking, view, completely, contr...   
157628        [ashamed, horrible, thing, put, talk, page]   
157629  [spitzer, umm, actual, article, prostitution, ...   
157630  [looks, like, actually, put, speedy, first, ve...   
157631  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  
0       explanation edits made username hardcore metal...  
1       daww match background colour seemingly stuck t...  
2       hey man really trying edit war guy constantly ...  
3       make real suggestion improvement wondered sect...  
4                           sir hero chance remember page  
...                                                   ...  
157627  second time asking view completely contradicts...  
157628               ashamed horrible thing put talk page  
157629  spitzer umm actual article prostitution ring c...  
157630  look like actually put speedy first version de...  
157631  really think understand came idea bad right aw...  

[157632 rows x 10 columns]

#### See the number of rows has reduced from 157669 to 157632 , so there was duplicantes in column lemmatization

## New column is made containing binary if it has offensive comment or not ; column called offensive

### made offensive column because it is more practical to have binary classification model that tells you this is offensive or non-offensive comment for real life applications ; also to help solving imbalace of the data

#### This column will be used in classification

In [63]:
# Create a new column with make value equal 1
df['offensive'] = 1

#codition is the row doesnot classify as toxic nor severe_toxic nor obscene nor threat nor insult nor identity_hate
condition = (df["toxic"]==0)&(df["severe_toxic"]==0)&(df["obscene"]==0)&(df["threat"]==0)&(df["insult"]==0)&(df["identity_hate"]==0)

# if it meet the condition and doesnot has offensive comment make the value equal 0
df.loc[condition, 'offensive'] = 0

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157627  ffe987279560d7ff  second time asking view completely contradicts...   
157628  ffea4adeee384e90               ashamed horrible thing put talk page   
157629  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157630  fff125370e4aaaf3  looks like actually put speedy first version d...   
157631  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157627      0             0        0       0       0              0   
157628      0             0        0       0       0              0   
157629      0             0        0       0       0              0   
157630      0             0        0       0       0              0   
157631      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157627  [second, time, asking, view, completely, contr...   
157628        [ashamed, horrible, thing, put, talk, page]   
157629  [spitzer, umm, actual, article, prostitution, ...   
157630  [looks, like, actually, put, speedy, first, ve...   
157631  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  offensive  
0       explanation edits made username hardcore metal...          0  
1       daww match background colour seemingly stuck t...          0  
2       hey man really trying edit war guy constantly ...          0  
3       make real suggestion improvement wondered sect...          0  
4                           sir hero chance remember page          0  
...                                                   ...        ...  
157627  second time asking view completely contradicts...          0  
157628               ashamed horrible thing put talk page          0  
157629  spitzer umm actual article prostitution ring c...          0  
157630  look like actually put speedy first version de...          0  
157631  really think understand came idea bad right aw...          0  

[157632 rows x 11 columns]

### A new column called offensive is made 

## Vocabulary Creation

### Make two dictionary to make each unique word with an unique index
- token_to_index : the word is the index
- index_to_token : the number is the index 
### There is another dictionary made 
- vocabulary

In [64]:
# function takes column name and data
def create_word_vocabulary_and_index(dataset,column_name):
    # initialize the set 
    vocabulary = set()
    # Start with [PAD] as index 0 in token_to_index 
    # because when make any padding which equal to zero the PAD holds the same value
    token_to_index = {'[PAD]': 0}
    # Start with 0 as index [PAD] in index_to_token
    index_to_token = {0: '[PAD]'}

    # Start the index = 1
    current_index = 1  
    # iterate over each row in data on a specific column
    for text in dataset[column_name]:
        # split data into word
        tokens = text.split()
        # iterate over each word
        for token in tokens:
            # if the word appears for the first time in token_to_index dictionary
            if token not in token_to_index:
                # token_to_index with index word = the current index
                token_to_index[token] = current_index
                # index_to_token with index current_index = the current word
                index_to_token[current_index] = token
                # add +1 to index
                current_index += 1
        # update the tokens in the vocabulary
        vocabulary.update(tokens)

    return token_to_index, index_to_token , vocabulary

### Apply function on data

In [65]:
# apply function on lemmatization column
token_to_index, index_to_token , vocabulary = create_word_vocabulary_and_index(df, 'lemmatization_column')

In [66]:
token_to_index

{'[PAD]': 0,
 'explanation': 1,
 'edits': 2,
 'made': 3,
 'username': 4,
 'hardcore': 5,
 'metallica': 6,
 'fan': 7,
 'reverted': 8,
 'vandalism': 9,
 'closure': 10,
 'gas': 11,
 'voted': 12,
 'new': 13,
 'york': 14,
 'doll': 15,
 'fac': 16,
 'please': 17,
 'remove': 18,
 'template': 19,
 'talk': 20,
 'page': 21,
 'since': 22,
 'retired': 23,
 'daww': 24,
 'match': 25,
 'background': 26,
 'colour': 27,
 'seemingly': 28,
 'stuck': 29,
 'thanks': 30,
 'january': 31,
 'utc': 32,
 'hey': 33,
 'man': 34,
 'really': 35,
 'trying': 36,
 'edit': 37,
 'war': 38,
 'guy': 39,
 'constantly': 40,
 'removing': 41,
 'relevant': 42,
 'information': 43,
 'talking': 44,
 'instead': 45,
 'seems': 46,
 'care': 47,
 'formatting': 48,
 'actual': 49,
 'info': 50,
 'make': 51,
 'real': 52,
 'suggestion': 53,
 'improvement': 54,
 'wondered': 55,
 'section': 56,
 'statistic': 57,
 'later': 58,
 'subsection': 59,
 'type': 60,
 'accident': 61,
 'think': 62,
 'reference': 63,
 'may': 64,
 'need': 65,
 'tidying': 6

In [67]:
index_to_token

{0: '[PAD]',
 1: 'explanation',
 2: 'edits',
 3: 'made',
 4: 'username',
 5: 'hardcore',
 6: 'metallica',
 7: 'fan',
 8: 'reverted',
 9: 'vandalism',
 10: 'closure',
 11: 'gas',
 12: 'voted',
 13: 'new',
 14: 'york',
 15: 'doll',
 16: 'fac',
 17: 'please',
 18: 'remove',
 19: 'template',
 20: 'talk',
 21: 'page',
 22: 'since',
 23: 'retired',
 24: 'daww',
 25: 'match',
 26: 'background',
 27: 'colour',
 28: 'seemingly',
 29: 'stuck',
 30: 'thanks',
 31: 'january',
 32: 'utc',
 33: 'hey',
 34: 'man',
 35: 'really',
 36: 'trying',
 37: 'edit',
 38: 'war',
 39: 'guy',
 40: 'constantly',
 41: 'removing',
 42: 'relevant',
 43: 'information',
 44: 'talking',
 45: 'instead',
 46: 'seems',
 47: 'care',
 48: 'formatting',
 49: 'actual',
 50: 'info',
 51: 'make',
 52: 'real',
 53: 'suggestion',
 54: 'improvement',
 55: 'wondered',
 56: 'section',
 57: 'statistic',
 58: 'later',
 59: 'subsection',
 60: 'type',
 61: 'accident',
 62: 'think',
 63: 'reference',
 64: 'may',
 65: 'need',
 66: 'tidying

In [68]:
vocabulary

{'ettentio',
 'agresion',
 'gambon',
 'napeolonic',
 'kokobo',
 'kyd',
 'seriouslyfolks',
 'userakraj',
 'pleny',
 'grounded',
 'raychaudhuri',
 'welladvised',
 'wulfsbergs',
 'reassessmenthalfway',
 'nowcoldplay',
 'dictorial',
 'tandy',
 'uniti',
 'skyisblue',
 'sarsa',
 'northholland',
 'kitaky',
 'bedevilment',
 'aquinas',
 'apprenticefan',
 'asslicking',
 'middleschoolers',
 'knowledgesharing',
 'familiarising',
 'lithgows',
 'tortuous',
 'inquiriesstyle',
 'optically',
 'conven',
 'imageheroess',
 'cooridinates',
 'fffaef',
 'talkwatts',
 'theni',
 'subalpine',
 'haiducs',
 'thisplease',
 'elitexc',
 'deletedreverted',
 'generous',
 'selfjustifications',
 'permeates',
 'fattened',
 'rius',
 'manics',
 'hisin',
 'vintner',
 'probable',
 'stsot',
 'conflictrussian',
 'respectthat',
 'mihail',
 'userwikifunusa',
 'beowulf',
 'mitochrondrial',
 'wfrwikipediastatistiques',
 'heighho',
 'cockonce',
 'rcs',
 'interessterd',
 'copilot',
 'selfpublicity',
 'resize',
 'gensammelten',
 'eli

### See the number of unique words in data

In [69]:
len(index_to_token)

203143

### Two dictionaries is successfully made gives each unique word in data an unique number

## Padding

### Function is to let the length of each row is the same to enter the transformer 

In [70]:
# function takes column name and data , max sequence lenght , dictionary token_to_index , padding is set = 0 as default
def tokens_to_indices_with_padding(tokens, max_seq_length, token_to_index, padding_value=0):
    # Convert tokens to indices
    # moves over each word in the text and check if it in the dictionary 
    # if yes , takes the corresponding index and put the word in the list
    indices = [token_to_index[token] for token in tokens if token in token_to_index]
    
    # Apply padding or truncation
    # subtract the lenght of indices list from the max sequence lenght
    # to determine the number of padding elements
    pad_length = max_seq_length - len(indices)

    # create a list and take the first "max_seq_length" elements from indices
    # and add padding * pad length   
    padded_indices = indices[:max_seq_length] + [padding_value] * pad_length

    return padded_indices

### function to count number of words in lemmatized data to put a number for variable max_seq_len

In [71]:
# function to count words
def count_words(text):
    # split a sentence to words
    words = text.split()
    # return the length of words
    return len(words)

# apply function on lemmatization column and make new column
df['max_word_count'] = df['lemmatization_column'].apply(count_words)

# calculate the mean of the new column
mean_max_word_count = df['max_word_count'].mean()

print(f"Mean of the maximum word counts across all rows: {mean_max_word_count}")

Mean of the maximum word counts across all rows: 33.59922477669509


### I can put to max_seq_len = 50 because it is near the mean of words

### Apply function on data

In [72]:
# Initialize the maximum sequence length
max_seq_length = 50  
# This lets each row has a the maximum sequence length of 50


# make a function to make padding around each row
def apply_padding(row):
    # Split lemmatization column into words
    lemmatized_tokens = row['lemmatization_column'].split()  
    # applies tokens_to_indices_with_padding function on each row
    padded_indices = tokens_to_indices_with_padding(lemmatized_tokens, max_seq_length, token_to_index)
    return padded_indices

# apply the function to make padding column
df['padding_column'] = df.apply(apply_padding, axis=1)

In [73]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  daww matches background colour seemingly stuck...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
157627  ffe987279560d7ff  second time asking view completely contradicts...   
157628  ffea4adeee384e90               ashamed horrible thing put talk page   
157629  ffee36eab5c267c9  spitzer umm actual article prostitution ring c...   
157630  fff125370e4aaaf3  looks like actually put speedy first version d...   
157631  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
157627      0             0        0       0       0              0   
157628      0             0        0       0       0              0   
157629      0             0        0       0       0              0   
157630      0             0        0       0       0              0   
157631      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [daww, matches, background, colour, seemingly,...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
157627  [second, time, asking, view, completely, contr...   
157628        [ashamed, horrible, thing, put, talk, page]   
157629  [spitzer, umm, actual, article, prostitution, ...   
157630  [looks, like, actually, put, speedy, first, ve...   
157631  [really, think, understand, came, idea, bad, r...   

                                     lemmatization_column  offensive  \
0       explanation edits made username hardcore metal...          0   
1       daww match background colour seemingly stuck t...          0   
2       hey man really trying edit war guy constantly ...          0   
3       make real suggestion improvement wondered sect...          0   
4                           sir hero chance remember page          0   
...                                                   ...        ...   
157627  second time asking view completely contradicts...          0   
157628               ashamed horrible thing put talk page          0   
157629  spitzer umm actual article prostitution ring c...          0   
157630  look like actually put speedy first version de...          0   
157631  really think understand came idea bad right aw...          0   

        max_word_count                                     padding_column  
0                   23  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...  
1                   10  [24, 25, 26, 27, 28, 29, 30, 20, 31, 32, 0, 0,...  
2                   21  [33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 4...  
3                   48  [51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 6...  
4                    5  [92, 93, 94, 95, 21, 0, 0, 0, 0, 0, 0, 0, 0, 0...  
...                ...                                                ...  
157627              23  [1204, 605, 2776, 469, 1103, 10162, 6302, 1153...  
157628        

### A new column is made called padded_column making the text data numbers not text to enter the Neural Networks

## The Transfomer

### The transformer is started from a paper called "attention is all you need" .
The transformer containes many classes like :

1.  Positional Encoding: Positional encoding adds information about the position of each token in the sequence to the embeddings, helping the Transformer model understand the sequential order of the input data.

2. Position-wise Feed-Forward Networks: After attention, each position's representation undergoes a position-wise feed-forward network, which applies a set of learnable transformations to capture non-linear patterns and relationships within the encoded information.

3.  Multi-Head Attention: This step in the Transformer model involves computing multiple attention heads in parallel, enabling the model to focus on different parts of the input sequence simultaneously, enhancing its ability to capture complex relationships.

4. Encoder: The encoder consists of multiple layers of Multi-Head Attention and Position-wise Feed-Forward Networks. Each layer incrementally refines the input representations, allowing the model to capture both local and global dependencies in the data.

5. Decoder: The decoder in a Transformer model is composed of several layers, each with its own Multi-Head Attention mechanism and Position-wise Feed-Forward Networks. These layers generate the output sequence step by step, attending to the encoder's output and previously generated tokens, facilitating sequence-to-sequence tasks like machine translation or text generation.

6. Transformer: The Transformer class is the combination of all classes with each other 

### Positional Encoding class

In [74]:
class PositionalEncoding(nn.Module):
    # Constructor for the Positional Encoding class 
    # that calls max sequence length and dimension of model
    def __init__(self, d_model, max_seq_len):
        super(PositionalEncoding, self).__init__()
        
        # Initialize a matrix to hold positional encodings
        # torch matrix of zeros  
        self.pe = torch.zeros(max_seq_len, d_model)
        
        # Create a tensor representing the positions within the sequence
        # make 2 dimensional array from zero to max_seq_len-1 
        position = torch.arange(0, max_seq_len, dtype=torch.float).unsqueeze(1)
        
        # makes the denumerator of the formula
        # make a tensor from to dmodel that moves by 2
        div_term = torch.exp(torch.arange(0, d_model, 2, dtype=torch.float) * (-np.log(10000.0) / d_model))
        
        # Calculate the sine and cosine values for each position and dimension
        # sin for even positions and cosine for odd indices
        self.pe[:, 0::2] = torch.sin(position * div_term) # even
        self.pe[:, 1::2] = torch.cos(position * div_term) # odd
        
        # Add an extra dimension to the positional encodings to move around all batch of sentences
        self.pe = self.pe.unsqueeze(0)

    # Forward method for the Positional Encoding class
    def forward(self, x):
        # Add positional encodings to the input tensor
        return x + self.pe[:, :x.size(1)]

### FeedForward class

In [75]:
class FeedForward(nn.Module):
    # Constructor for the FeedForward class
    # that calls dimension of model and feedforward dimension
    def __init__(self, d_model, d_ff):
        super(FeedForward, self).__init__()
        
        # make a fully connected layer input nodes = dimension of model 
        # output nodes = feedforward dimension
        self.fc1 = nn.Linear(d_model, d_ff)
        
        # make a fully connected layer input nodes = feedforward dimension
        # output nodes = dimension of model
        self.fc2 = nn.Linear(d_ff, d_model)

    # Forward method for the FeedForward class
    def forward(self, x):
        # add Relu as activation function for fc1
        x = F.relu(self.fc1(x))
        
        # no activation function on fc2
        x = self.fc2(x)
        
        return x

### Multi-Head Attention class

In [76]:
class MultiHeadAttention(nn.Module):
    # Constructor for the Multi-Head Attention class 
    # that calls dimension of model and number of heads
    def __init__(self, d_model, n_heads):
        super(MultiHeadAttention, self).__init__()

        # Initialize dimension of model 
        self.d_model = d_model 

        # Initialize Number of  heads
        self.n_heads = n_heads 

        # Initialize the dimension of each attension head
        self.head_dim = d_model // n_heads  

        # Linear transformations for Query, Key, and Value
        # W_Q is for query that that will have a of shape (seq , d_model)
        self.W_Q = nn.Linear(d_model, d_model)

        # W_K is for key that will have a of shape (seq , d_model)
        self.W_K = nn.Linear(d_model, d_model)

        # W_V is for value that will have a of shape (seq , d_model)
        self.W_V = nn.Linear(d_model, d_model)

        # Output and will make the output shape (seq , d_model)
        self.W_O = nn.Linear(d_model, d_model)

    def forward(self, Q, K, V, mask=None):
        # Q, K, and V are the Query, Key, and Value tensors
        
        # Get the batch size from the input tensor of Query
        batch_size = Q.shape[0]

        # Project the input Query, Key, and Value tensors
        # multiply Q by w_Q to get output shape (seq , d_model)
        Q = self.W_Q(Q)
        # multiply K by w_K to get output shape (seq , d_model)
        K = self.W_K(K)
        # multiply V by w_V to get output shape (seq , d_model)
        V = self.W_V(V)

        # Divide the Query Q matrix to smaller matrices to give each smaller matrix to a different matrix
        Q = Q.view(batch_size, -1, self.n_heads, self.head_dim)
        # Divide the Key K matrix to smaller matrices to give each smaller matrix to a different matrix
        K = K.view(batch_size, -1, self.n_heads, self.head_dim)
        # Divide the value V matrix to smaller matrices to give each smaller matrix to a different matrix
        V = V.view(batch_size, -1, self.n_heads, self.head_dim)

        # make the matrix multiplication on Query and Key for each head
        # bqhd represents the dimensions of tensor Query in the form of (batch size, sequence length, number of attention heads, dimension of key vectors)
        # bkhd represents the dimensions of tensor Key in the form of (batch size, sequence length, number of attention heads, dimension of key vectors)
        # bhqk represents the output dimensions in the form of (batch size, number of attention heads, query sequence length, key sequence length)
        QK = torch.einsum('bqhd,bkhd->bhqk', Q, K)

        # get the attention scores by dividing the QK by square root of self.head_dim
        scores = QK / torch.sqrt(torch.tensor(self.head_dim).float())

        # Apply the attention mask if not set to None 
        if mask is not None:
            # make the mask match the dimension of the scores tensor
            mask = mask.unsqueeze(1).unsqueeze(2)
            # replace where the mask = zero by large negative number
            scores.masked_fill_(mask == 0, -1e9)

        # make softmax to get attention weights
        attention_weights = F.softmax(scores, dim=-1)

        # Compute matrix multiplication of attention_weights and Values
        # bhqv represents the dimensions of the attention_weights tensor in the form of (batch size, number of attention heads, query sequence length, value dimension)
        # bvhd represents the dimensions of the V tensor in the form of (batch size, sequence length, number of attention heads, dimension of value vectors)
        # bqhd represents the output dimensions in the form of (batch size, query sequence length, number of attention heads, dimension of value vectors)
        out = torch.einsum('bhqv,bvhd->bqhd', attention_weights, V)

        # reshapes the tensor to combine the num_heads and head_dimension
        out = out.reshape(batch_size, -1, self.d_model)

        # multiply W_O by out reshaped 
        # The result is a tensor of shape (batch_size, sequence_length)
        out = self.W_O(out)

        return out

### Encoder Layer Class

In [77]:
class EncoderLayer(nn.Module):
    # Constructor for the Encoder Layer class 
    # that calls dimension of model and number of heads and feedforward dimension
    def __init__(self, d_model, n_heads, d_ff):
        super(EncoderLayer, self).__init__()
        
        # call Multi-Head Attention class
        self.self_attention = MultiHeadAttention(d_model, n_heads)
        
        # call Feed-Forward class
        self.feed_forward = FeedForward(d_model, d_ff)
        
        # Initialize Layer Normalization for the dimension of model
        self.norm1 = nn.LayerNorm(d_model)
        
        # Initialize Layer Normalization for the dimension of model
        self.norm2 = nn.LayerNorm(d_model)

    def forward(self, x, mask):
        # Multi-Head Self-Attention
        attention_output = self.self_attention(x, x, x, mask)
        
        # Residual Connection for self-attention
        x = x + attention_output
        # Layer Normalization for self-attention
        x = self.norm1(x)
        
        # Feed-Forward Neural Network
        feed_forward_output = self.feed_forward(x)
        
        # Residual Connection for feed-forward
        x = x + feed_forward_output
        # Layer Normalization for feed-forward
        x = self.norm2(x)
        
        return x

### Transformer Encoder Class

In [78]:
class TransformerEncoder(nn.Module):
    # Constructor for the Transfomer Encoder class 
    # that calls dimension of model and number of heads and feedforward dimension and number of layers
    def __init__(self, num_layers, d_model, n_heads, d_ff):
        super(TransformerEncoder, self).__init__()

        # Create a list of number encoder layers 
        self.layers = nn.ModuleList([EncoderLayer(d_model, n_heads, d_ff) for _ in range(num_layers)])
    
    def forward(self, x, mask):
        # for the number of layers 
        for layer in self.layers:
            # make the input x which has shape of (batch_size, seq_len, d_model)
            # mask to hide padded elements
            x = layer(x, mask)
        return x


### Transfomer Class

In [79]:
class Transformer(nn.Module):
    # Constructor for the Transfomer class 
    # that calls dimension of model and number of heads and feedforward dimension and number of layers
    # and size of vocabulary and max sequence length and number of output classes
    def __init__(self, num_layers, d_model, n_heads, d_ff, input_vocab_size, max_seq_len, num_classes):
        super(Transformer, self).__init__()

        # make the embedding layer
        self.embedding = nn.Embedding(input_vocab_size, d_model)

        # call Positional Encoding class
        self.positional_encoding = PositionalEncoding(d_model, max_seq_len)

        # call Transformer Encoder class
        self.encoder = TransformerEncoder(num_layers, d_model, n_heads, d_ff)

        # Initialize a fully conected layer 
        self.fc = nn.Linear(d_model, num_classes)

    def forward(self, x, mask):
        # Pass input tokens into the embedding layer
        x = self.embedding(x)

        # Apply positional encoding to the embedded output
        x = self.positional_encoding(x)

        # let the output of positional encoding pass through the encoder layer
        x = self.encoder(x, mask)

        # make average pooling on sequence length dimension on output of the encoder
        x = x.mean(dim=1)  

        # pass the output of average pooling through the fully connencted layer
        x = self.fc(x)

        return x

### Didn't make an decoder layer because my goal to put a category to an input, not to generate a sequence-to-sequence as decoder layer does that is used for machine translation ; and also decoder makes unnecessary complexity 

### Training function

In [80]:
def train(model, train_loader, valid_loader, criterion, optimizer, num_epochs, patience):
    # make the model in training mode
    model.train()

    # Initialize the best validation loss to a large value
    # that is for early stopping to not make overfitting
    best_valid_loss = float('inf')  

    # Initialize a counter for consecutive epochs with no improvement
    # also for early stopping to not make overfitting 
    consecutive_no_improvement = 0  
    
    # for loop for the number of epoches
    for epoch in range(num_epochs):
        # Initialize the loss = zero
        running_loss = 0.0
        # Initialize the correct_predictions = zero
        correct_predictions = 0
        # Initialize the total_samples = zero
        total_samples = 0
        
        # Loop in each batch in train data
        for batch_idx, (inputs, targets) in enumerate(train_loader, 1):
            # make zero to the grad 
            optimizer.zero_grad()  
             # put mask to input not equal to zero
            mask = (inputs != 0) 
            # Forward pass through the model
            outputs = model(inputs, mask)  
            # Calculate the loss
            loss = criterion(outputs, targets)  
            # Backprop 
            loss.backward()  
             # Update the weights
            optimizer.step() 
            # add batch loss to epoch loss
            running_loss += loss.item()  
            
            # store the prediction of model
            _, predicted = torch.max(outputs, 1)
            # if predicted = target this correct prediction add it to the correct prediction
            correct_predictions += (predicted == targets).sum().item()
            # add the number of samples
            total_samples += targets.size(0)
            
            # Print progress information the number of epoch done / the total number of epochs and number of batchs done / the total number of batchs
            print(f"\rEpoch {epoch + 1}/{num_epochs}, Batch {batch_idx}/{len(train_loader)}", end='', flush=True)

        # Calculate the average loss for the epoch
        epoch_loss = running_loss / len(train_loader)  
        # Calculate the epoch accuracy
        epoch_accuracy = 100 * correct_predictions / total_samples  
        
        
       # make the model in evaluation mode for Validation
        model.eval()  
        # Initialize the loss = zero
        valid_loss = 0.0
        # Initialize the correct_predictions = zero
        valid_correct = 0
        # Initialize the total_samples = zero
        valid_total = 0
        
        # donot make gradient 
        with torch.no_grad():
            # Loop in each batch in validation data
            for valid_inputs, valid_targets in valid_loader:
                # put mask to input not equal to zero
                valid_mask = (valid_inputs != 0)
                # Forward pass through the model
                valid_outputs = model(valid_inputs, valid_mask) 
                # Calculate validation loss
                valid_loss_batch = criterion(valid_outputs, valid_targets)  
                # add batch loss to epoch loss
                valid_loss += valid_loss_batch.item()  
                
                # store the prediction of model
                _, valid_predicted = torch.max(valid_outputs, 1)
                # add the number of samples
                valid_total += valid_targets.size(0)
                # if predicted = target this correct prediction add it to the correct prediction
                valid_correct += (valid_predicted == valid_targets).sum().item()

        # Calculate the epoch accuracy
        valid_accuracy = 100 * valid_correct / valid_total 
        # Calculate the average loss for the epoch
        avg_valid_loss = valid_loss / len(valid_loader)  
        
        # print train loss and acc for validation and train data for each epoch
        print(f"\rEpoch {epoch + 1}/{num_epochs}, Train Loss: {epoch_loss:.4f}, Train Accuracy: {epoch_accuracy:.2f}%, " \
              f"Valid Loss: {avg_valid_loss:.4f}, Valid Accuracy: {valid_accuracy:.2f}%")
        
        # for Early stopping 
        # Check if the validation loss has improved
        if avg_valid_loss < best_valid_loss:
            best_valid_loss = avg_valid_loss
            consecutive_no_improvement = 0
        else:
            # if loss increased add one to the no improvement variable
            consecutive_no_improvement += 1
        
        # If there is no improvement variable exceeded the patient
        if consecutive_no_improvement >= patience:
            print(f"Early stopping after {epoch + 1} epochs with no improvement.")
            # stop the function
            break
        
        # make the model in training mode
        model.train()  

### test function

In [81]:
def test(model, test_loader, criterion):
    # make the model in evaluation mode
    model.eval() 
    # Initialize the correct_predictions = zero
    correct = 0  
    # Initialize the total_samples = zero
    total = 0
    # Initialize the loss = zero
    test_loss = 0.0  
    # array for all predicted values
    all_predicted = []
    # array for all true values
    all_targets = []

    # donot make gradient 
    with torch.no_grad():
        # Loop in each batch in test data
        for inputs, targets in test_loader:
            # put mask to input not equal to zero
            mask = (inputs != 0) 
            # Forward pass through the model
            outputs = model(inputs, mask) 
            # Calculate the loss
            loss = criterion(outputs, targets)
            # add batch loss to loss
            test_loss += loss.item()  
            
            # store the prediction of model
            _, predicted = torch.max(outputs, 1)
            # add the number of samples
            total += targets.size(0)
            # if predicted = target this correct prediction add it to the correct prediction
            correct += (predicted == targets).sum().item()

            # add predicted values
            all_predicted.extend(predicted.tolist())
            # add true values
            all_targets.extend(targets.tolist())
    
    # Calculate test accuracy
    accuracy = 100 * correct / total  
    # Calculate the average test loss
    avg_test_loss = test_loss / len(test_loader)  
    
    # print test loss and acc 
    print(f"Test Loss: {avg_test_loss:.4f}, Test Accuracy: {accuracy:.2f}%")

    # print the classification report
    print("\nClassification Report:")
    report = classification_report(all_targets, all_predicted)
    print(report)

### Initialize train , validation , and test data

In [82]:
# convert padding column to numpy array 
padding_array = np.array(df['padding_column'].tolist())

# convert padded array to tensor
X = torch.from_numpy(padding_array).long()

# convert offensive column that has classification to tensor
y = torch.from_numpy(df['offensive'].values).long()

In [83]:
# Split the data into train, test 
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)
# split test data into test and validation
X_valid, X_test, y_valid, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

In [84]:
# Initialize batch size
batch_size = 64

# make tensor train data 
train_dataset = TensorDataset(X_train, y_train)
# make train data iterable and make batchs
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

# make tensor validation data 
valid_dataset = TensorDataset(X_valid, y_valid)
# make validation data iterable and make batchs
valid_loader = DataLoader(valid_dataset, batch_size=batch_size)

# make tensor test data 
test_dataset = TensorDataset(X_test, y_test)
# make test data iterable and make batchs
test_loader = DataLoader(test_dataset, batch_size=batch_size)

### Define hyperparameters

In [85]:
# number of layers
num_layers = 6

# dimension of the model
d_model = 512

# number of attension heads
n_heads = 8

# feedforward dimension 
d_ff = 2048 

# vocab size = length of dictionary token_to_index or index_to_token they have same length
input_vocab_size = len(token_to_index)  

# make max sequence length as above
max_seq_len = 50  

# good learning rate not small nor big
lr = 0.0001

# for binary classification so 2 classes
num_classes = 2  

#### I tried different hyperparameters but those get the best loss and accuracy

### Initialize model , optimizer , criterion

In [86]:
model = Transformer(num_layers, d_model, n_heads, d_ff, input_vocab_size, max_seq_len, num_classes)
# cross entropy loss function
criterion = nn.CrossEntropyLoss()
# adam optimizer
optimizer = optim.Adam(model.parameters(), lr=lr)

### Initialize the number of epoches , patience for early stopping 
### Apply of training function

In [87]:
# define number of epoches
num_epochs = 8
# define number of patience , which is the parameter for early stopping  if validation loss increased by this number will stop training
patience = 3

# train the model
train(model, train_loader, valid_loader, criterion, optimizer, num_epochs, patience)

Epoch 1/8, Train Loss: 0.1688, Train Accuracy: 94.64%, Valid Loss: 0.1358, Valid Accuracy: 95.35%
Epoch 2/8, Train Loss: 0.1145, Train Accuracy: 95.92%, Valid Loss: 0.1241, Valid Accuracy: 95.77%
Epoch 3/8, Train Loss: 0.0827, Train Accuracy: 96.96%, Valid Loss: 0.1420, Valid Accuracy: 95.60%
Epoch 4/8, Train Loss: 0.0465, Train Accuracy: 98.36%, Valid Loss: 0.1852, Valid Accuracy: 95.23%
Epoch 5/8, Train Loss: 0.0283, Train Accuracy: 98.98%, Valid Loss: 0.1808, Valid Accuracy: 95.43%
Early stopping after 5 epochs with no improvement.


#### the training accuracy increased in each epoch , validation loss was best in second epoch then increased which will lead to overfitting so I put early stopping to not lead to overfitting

### Apply of test function

In [88]:
# test the model
test(model, test_loader, criterion)

Test Loss: 0.1847, Test Accuracy: 95.43%

Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.98      0.97     28310
           1       0.84      0.69      0.75      3217

    accuracy                           0.95     31527
   macro avg       0.90      0.84      0.86     31527
weighted avg       0.95      0.95      0.95     31527



#### explaination of the output
- Precision : measure of how many of the predicted positive instances were actually positive
- Recall : measure of how many of the actual positive instances were correctly predicted as positive
- F1-Score : harmonic mean of precision and recall
- Support : number of instances in the true dataset that belong to each class.
- Accuracy : the accuracy of model which is 95 %
- Macro Avg : These values are the unweighted average of precision, recall, and F1-score for all classes 
- Sample Avg : These values are similar to the micro average but consider each sample equally.

#### The model performers well with 95% accuracy 
#### percision and recall on class 0 indicates that perdictes this class very well ; also it perdictes class 1 good but not as good as class 0